File: bin/clone.ts
import run from './run';
import { URL } from 'url';

const clone = (urlString:string, location='~/src/') => {
  const url = new URL(urlString);

  const folderPath = location + url.pathname.split('/')[1];
  const repoName = url.pathname.split('/')[2].replace('.git', '');
  const repoPath = `${folderPath}/${repoName}`;

  try {
    run(`mkdir -p ${folderPath} && cd ${folderPath} && git clone ${url.href}`);
    console.log(`Successfully cloned ${repoName} into ${repoPath}`);
  } catch (err) {
    console.log('quick: unable to clone repository');
  }
  console.log(`\ncd ${repoPath}`);
}

export default clone;

File: lib/clone.rb
# typed: strict
# frozen_string_literal: true

require 'fileutils'
require 'uri'

module Quick
  class Clone
    class << self
      extend T::Sig

      sig { params(url: String).void }
      def clone_repository(url)
        uri = URI.parse(url)
        path_segments = uri.path.split('/')
        github_username = path_segments[1]
        github_reponame = path_segments[2].gsub('.git', '')

        dest_dir = File.expand_path("~/src/#{github_username}/#{github_reponame}")

        if Dir.exist?(dest_dir) && !Dir.empty?(dest_dir)
          puts "Error: Destination path '#{dest_dir}' already exists and is not an empty directory."
          return
        end

        FileUtils.mkdir_p(dest_dir)

        puts "Cloning into #{dest_dir}..."
        puts "cd '#{dest_dir}'" if system("git clone #{url} #{dest_dir}")
      end
    end
  end
end
